File: backend/src/main/java/pl/strava/analizator/domain/workout/WorkoutRunner.java
package pl.strava.analizator.domain.workout;

import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

import pl.strava.analizator.domain.model.WorkoutExecution;
import pl.strava.analizator.domain.model.WorkoutExecutionStatus;
import pl.strava.analizator.domain.model.WorkoutStep;

/**
 * Pure workout state machine. Wall-clock instants are persisted, so state can be
 * reconstructed after a process or device restart. A UI timer is only a repaint trigger.
 */
public final class WorkoutRunner {

    private static final int MIN_INTENSITY_ADJUSTMENT = -50;
    private static final int MAX_INTENSITY_ADJUSTMENT = 50;

    private WorkoutRunner() {
    }

    public static WorkoutExecution start(WorkoutExecution state, Instant now) {
        requireStatus(state, WorkoutExecutionStatus.READY);
        requireSteps(state);
        return state.toBuilder()
                .status(WorkoutExecutionStatus.RUNNING)
                .startedAt(state.getStartedAt() != null ? state.getStartedAt() : now)
                .runningSince(now)
                .stateVersion(state.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    public static WorkoutExecution reconcile(WorkoutExecution state, Instant now) {
        if (state.getStatus() != WorkoutExecutionStatus.RUNNING || state.getRunningSince() == null) {
            return state;
        }
        long availableMs = Math.max(0, Duration.between(state.getRunningSince(), now).toMillis());
        if (availableMs == 0) {
            return state;
        }

        List<WorkoutStep> steps = safeSteps(state);
        int stepIndex = state.getCurrentStepIndex();
        long stepElapsed = state.getStepElapsedMs();
        long workoutElapsed = state.getWorkoutElapsedMs();
        Instant cursor = state.getRunningSince();

        while (availableMs > 0 && stepIndex < steps.size()) {
            Long durationMs = durationMs(steps.get(stepIndex));
            if (durationMs == null) {
                stepElapsed += availableMs;
                workoutElapsed += availableMs;
                cursor = cursor.plusMillis(availableMs);
                availableMs = 0;
                break;
            }

            long remainingInStep = Math.max(0, durationMs - stepElapsed);
            long consumed = Math.min(availableMs, remainingInStep);
            stepElapsed += consumed;
            workoutElapsed += consumed;
            cursor = cursor.plusMillis(consumed);
            availableMs -= consumed;

            if (stepElapsed >= durationMs) {
                stepIndex++;
                stepElapsed = 0;
            }
            if (remainingInStep == 0 && consumed == 0 && stepIndex < steps.size()) {
                continue;
            }
        }

        if (stepIndex >= steps.size()) {
            return state.toBuilder()
                    .status(WorkoutExecutionStatus.COMPLETED)
                    .currentStepIndex(Math.max(0, steps.size() - 1))
                    .stepElapsedMs(steps.isEmpty() ? 0 : valueOrZero(durationMs(steps.getLast())))
                    .workoutElapsedMs(workoutElapsed)
                    .runningSince(null)
                    .finishedAt(cursor)
                    .stateVersion(state.getStateVersion() + 1)
                    .updatedAt(cursor)
                    .build();
        }

        return state.toBuilder()
                .currentStepIndex(stepIndex)
                .stepElapsedMs(stepElapsed)
                .workoutElapsedMs(workoutElapsed)
                .runningSince(now)
                .stateVersion(state.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    public static WorkoutExecution pause(WorkoutExecution state, Instant now) {
        requireStatus(state, WorkoutExecutionStatus.RUNNING);
        WorkoutExecution reconciled = reconcile(state, now);
        if (reconciled.getStatus() == WorkoutExecutionStatus.COMPLETED) {
            return reconciled;
        }
        return reconciled.toBuilder()
                .status(WorkoutExecutionStatus.PAUSED)
                .runningSince(null)
                .stateVersion(reconciled.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    public static WorkoutExecution resume(WorkoutExecution state, Instant now) {
        requireStatus(state, WorkoutExecutionStatus.PAUSED);
        return state.toBuilder()
                .status(WorkoutExecutionStatus.RUNNING)
                .runningSince(now)
                .stateVersion(state.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    public static WorkoutExecution skipStep(WorkoutExecution state, Instant now) {
        WorkoutExecution reconciled = reconcileForAction(state, now);
        int skipped = reconciled.getCurrentStepIndex();
        List<Integer> skippedSteps = append(reconciled.getSkippedStepIndexes(), skipped);
        int next = skipped + 1;
        if (next >= safeSteps(reconciled).size()) {
            return complete(reconciled, now).toBuilder()
                    .skippedStepIndexes(skippedSteps)
                    .build();
        }
        return reconciled.toBuilder()
                .currentStepIndex(next)
                .stepElapsedMs(0)
                .runningSince(reconciled.getStatus() == WorkoutExecutionStatus.RUNNING ? now : null)
                .skippedStepIndexes(skippedSteps)
                .stateVersion(reconciled.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    public static WorkoutExecution previousStep(WorkoutExecution state, Instant now) {
        WorkoutExecution reconciled = reconcileForAction(state, now);
        int previous = Math.max(0, reconciled.getCurrentStepIndex() - 1);
        return reconciled.toBuilder()
                .currentStepIndex(previous)
                .stepElapsedMs(0)
                .runningSince(reconciled.getStatus() == WorkoutExecutionStatus.RUNNING ? now : null)
                .repeatedStepIndexes(append(reconciled.getRepeatedStepIndexes(), previous))
                .stateVersion(reconciled.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    public static WorkoutExecution changeIntensity(WorkoutExecution state, int deltaPct, Instant now) {
        WorkoutExecution reconciled = reconcileForAction(state, now);
        int adjusted = Math.max(MIN_INTENSITY_ADJUSTMENT,
                Math.min(MAX_INTENSITY_ADJUSTMENT, reconciled.getIntensityAdjustmentPct() + deltaPct));
        return reconciled.toBuilder()
                .intensityAdjustmentPct(adjusted)
                .runningSince(reconciled.getStatus() == WorkoutExecutionStatus.RUNNING ? now : null)
                .stateVersion(reconciled.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    public static WorkoutExecution complete(WorkoutExecution state, Instant now) {
        WorkoutExecution reconciled = reconcileForAction(state, now);
        if (reconciled.getStatus() == WorkoutExecutionStatus.COMPLETED) {
            return reconciled;
        }
        return reconciled.toBuilder()
                .status(WorkoutExecutionStatus.COMPLETED)
                .runningSince(null)
                .finishedAt(now)
                .stateVersion(reconciled.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    public static WorkoutExecution abort(WorkoutExecution state, Instant now) {
        WorkoutExecution reconciled = reconcileForAction(state, now);
        if (reconciled.getStatus() == WorkoutExecutionStatus.COMPLETED) {
            throw new IllegalStateException("Completed workout cannot be aborted");
        }
        return reconciled.toBuilder()
                .status(WorkoutExecutionStatus.ABORTED)
                .runningSince(null)
                .finishedAt(now)
                .stateVersion(reconciled.getStateVersion() + 1)
                .updatedAt(now)
                .build();
    }

    private static WorkoutExecution reconcileForAction(WorkoutExecution state, Instant now) {
        if (state.getStatus() != WorkoutExecutionStatus.RUNNING
                && state.getStatus() != WorkoutExecutionStatus.PAUSED) {
            throw new IllegalStateException("Workout action is not allowed from " + state.getStatus());
        }
        return reconcile(state, now);
    }

    private static void requireStatus(WorkoutExecution state, WorkoutExecutionStatus expected) {
        if (state.getStatus() != expected) {
            throw new IllegalStateException("Expected " + expected + " but was " + state.getStatus());
        }
    }

    private static void requireSteps(WorkoutExecution state) {
        if (safeSteps(state).isEmpty()) {
            throw new IllegalStateException("Workout requires at least one step");
        }
    }

    private static List<WorkoutStep> safeSteps(WorkoutExecution state) {
        return state.getStepsSnapshot() != null ? state.getStepsSnapshot() : List.of();
    }

    private static Long durationMs(WorkoutStep step) {
        if (step.getDurationSec() == null
                || "LAP_BUTTON".equalsIgnoreCase(step.getDurationType())
                || "OPEN".equalsIgnoreCase(step.getDurationType())) {
            return null;
        }
        return Math.max(0, step.getDurationSec()) * 1_000L;
    }

    private static long valueOrZero(Long value) {
        return value != null ? value : 0;
    }

    private static List<Integer> append(List<Integer> values, int value) {
        List<Integer> result = new ArrayList<>(values != null ? values : List.of());
        result.add(value);
        return List.copyOf(result);
    }
}

File: frontend/src/features/workout/workoutRunner.ts
import type { WorkoutExecution, WorkoutStep } from '@/types/training';

export type RunnerAction =
  | { type: 'RECONCILE'; nowMs: number }
  | { type: 'PAUSE'; nowMs: number }
  | { type: 'RESUME'; nowMs: number }
  | { type: 'SKIP'; nowMs: number }
  | { type: 'PREVIOUS'; nowMs: number }
  | { type: 'INTENSITY'; nowMs: number; deltaPct: number }
  | { type: 'COMPLETE'; nowMs: number }
  | { type: 'ABORT'; nowMs: number };

function durationMs(step: WorkoutStep): number | null {
  if (step.durationSec == null || step.durationType === 'OPEN' || step.durationType === 'LAP_BUTTON') {
    return null;
  }
  return Math.max(0, step.durationSec * 1000);
}

function versioned(state: WorkoutExecution, nowMs: number): WorkoutExecution {
  return { ...state, stateVersion: state.stateVersion + 1, updatedAt: new Date(nowMs).toISOString() };
}

export function reconcileWorkout(state: WorkoutExecution, nowMs: number): WorkoutExecution {
  if (state.status !== 'RUNNING' || state.runningSince == null) return state;
  let available = Math.max(0, nowMs - Date.parse(state.runningSince));
  if (available === 0) return state;
  let stepIndex = state.currentStepIndex;
  let stepElapsed = state.stepElapsedMs;
  let workoutElapsed = state.workoutElapsedMs;
  let cursor = Date.parse(state.runningSince);

  while (available > 0 && stepIndex < state.stepsSnapshot.length) {
    const currentStep = state.stepsSnapshot[stepIndex];
    if (!currentStep) break;
    const stepDuration = durationMs(currentStep);
    if (stepDuration == null) {
      stepElapsed += available;
      workoutElapsed += available;
      cursor += available;
      available = 0;
      break;
    }
    const remaining = Math.max(0, stepDuration - stepElapsed);
    const consumed = Math.min(available, remaining);
    stepElapsed += consumed;
    workoutElapsed += consumed;
    cursor += consumed;
    available -= consumed;
    if (stepElapsed >= stepDuration) {
      stepIndex += 1;
      stepElapsed = 0;
    }
  }

  if (stepIndex >= state.stepsSnapshot.length) {
    const last = state.stepsSnapshot[state.stepsSnapshot.length - 1];
    return versioned({
      ...state,
      status: 'COMPLETED',
      currentStepIndex: Math.max(0, state.stepsSnapshot.length - 1),
      stepElapsedMs: last ? (durationMs(last) ?? stepElapsed) : 0,
      workoutElapsedMs: workoutElapsed,
      runningSince: null,
      finishedAt: new Date(cursor).toISOString(),
    }, cursor);
  }
  return versioned({
    ...state,
    currentStepIndex: stepIndex,
    stepElapsedMs: stepElapsed,
    workoutElapsedMs: workoutElapsed,
    runningSince: new Date(nowMs).toISOString(),
  }, nowMs);
}

export function workoutReducer(state: WorkoutExecution, action: RunnerAction): WorkoutExecution {
  const reconciled = action.type === 'RESUME' ? state : reconcileWorkout(state, action.nowMs);
  const now = new Date(action.nowMs).toISOString();
  switch (action.type) {
    case 'RECONCILE':
      return reconciled;
    case 'PAUSE':
      if (reconciled.status === 'COMPLETED') return reconciled;
      if (reconciled.status !== 'RUNNING') throw new Error('Pauza jest dostępna tylko podczas treningu');
      return versioned({ ...reconciled, status: 'PAUSED', runningSince: null }, action.nowMs);
    case 'RESUME':
      if (state.status !== 'PAUSED') throw new Error('Można wznowić tylko wstrzymany trening');
      return versioned({ ...state, status: 'RUNNING', runningSince: now }, action.nowMs);
    case 'SKIP': {
      if (reconciled.status !== 'RUNNING' && reconciled.status !== 'PAUSED') throw new Error('Trening nie jest aktywny');
      const skipped = [...reconciled.skippedStepIndexes, reconciled.currentStepIndex];
      if (reconciled.currentStepIndex + 1 >= reconciled.stepsSnapshot.length) {
        return versioned({ ...reconciled, status: 'COMPLETED', finishedAt: now, runningSince: null, skippedStepIndexes: skipped }, action.nowMs);
      }
      return versioned({ ...reconciled, currentStepIndex: reconciled.currentStepIndex + 1, stepElapsedMs: 0, skippedStepIndexes: skipped, runningSince: reconciled.status === 'RUNNING' ? now : null }, action.nowMs);
    }
    case 'PREVIOUS': {
      if (reconciled.status !== 'RUNNING' && reconciled.status !== 'PAUSED') throw new Error('Trening nie jest aktywny');
      const previous = Math.max(0, reconciled.currentStepIndex - 1);
      return versioned({ ...reconciled, currentStepIndex: previous, stepElapsedMs: 0, repeatedStepIndexes: [...reconciled.repeatedStepIndexes, previous], runningSince: reconciled.status === 'RUNNING' ? now : null }, action.nowMs);
    }
    case 'INTENSITY':
      return versioned({ ...reconciled, intensityAdjustmentPct: Math.max(-50, Math.min(50, reconciled.intensityAdjustmentPct + action.deltaPct)), runningSince: reconciled.status === 'RUNNING' ? now : null }, action.nowMs);
    case 'COMPLETE':
      return versioned({ ...reconciled, status: 'COMPLETED', finishedAt: now, runningSince: null }, action.nowMs);
    case 'ABORT':
      if (reconciled.status === 'COMPLETED') throw new Error('Ukończonego treningu nie można przerwać');
      return versioned({ ...reconciled, status: 'ABORTED', finishedAt: now, runningSince: null }, action.nowMs);
  }
}

export function createMonotonicClock(
  wallNow: () => number = Date.now,
  monotonicNow: () => number = () => performance.now(),
) {
  const wallOrigin = wallNow();
  const monotonicOrigin = monotonicNow();
  return () => wallOrigin + Math.max(0, monotonicNow() - monotonicOrigin);
}

export function totalWorkoutDurationMs(steps: WorkoutStep[]): number {
  return steps.reduce((total, step) => total + (durationMs(step) ?? 0), 0);
}
